casper/src/main/scala/coop/rchain/casper/blocks/proposer/Proposer.scala

```diff
@@ -128,10 +133,111 @@ object Proposer {
+          val msgMap = dag.dagMessageState.msgMap
+          parentHashes.flatMap(msgMap(_).seen) -- preState.fringe.flatMap(msgMap(_).seen)
+        }
+        nothingToFinalize = conflictSet.toList
+          .traverse(BlockStore[F].getUnsafe)
+          .map(_.forall { b =>
+            b.state.systemDeploys.isEmpty && b.state.deploys.isEmpty
+          })
+        waitingForSupermajorityToAttest = {
+          val newlySeen = creatorsLatestOpt
+            .map { prev =>
+              prev.justifications.flatMap(dag.dagMessageState.msgMap(_).seen) --
+                parentHashes.flatMap(dag.dagMessageState.msgMap(_).seen)
+            }
+            .getOrElse(Set())
+          newlySeen.toList.traverse(BlockStore[F].getUnsafe).map { newBlocks =>
+            val newStateTransition =
+              newBlocks.exists(b => b.state.systemDeploys.nonEmpty || b.state.deploys.nonEmpty)
+            val attestationStake =
+              preStateBonds.filterKeys(newBlocks.map(_.sender).toSet).values.toList.sum
+            !newStateTransition && attestationStake * 3 < preStateBonds.values.toList.sum * 2
+          }
+        }
+        suppressAttestation <- nothingToFinalize ||^ waitingForSupermajorityToAttest
+//        // push dummy deploy if needed
+//        p <- BlockDagStorage[F].pooledDeploys
+//        _ <- dummyDeployOpt
+//              .traverse {
+//                case (privateKey, term) =>
+//                  val deployData = ConstructDeploy.sourceDeployNow(
+//                    source = term,
+//                    sec = privateKey,
+//                    vabn = nextBlockNum - 1,
+//                    shardId = shardId
+//                  )
+//                  BlockDagStorage[F].addDeploy(deployData)
+//              }
+//              .whenA(p.isEmpty)
+        // user deploys
+        pooled <- BlockDagStorage[F].pooledDeploys
+        pooledOk <- pooled.toList
+                     .filterA {
+                       case (id, d) =>
+                         val future       = d.data.validAfterBlockNumber > nextBlockNum
+                         val expired      = d.data.validAfterBlockNumber < nextBlockNum - MultiParentCasper.deployLifespan
+                         val replayAttack = BlockDagStorage[F].lookupByDeployId(id).map(_.nonEmpty)
+                         (future.pure ||^ expired.pure ||^ replayAttack).not
+                     }
+                     .map(_.map(_._1))
+        deploys <- {
+          val dummy = dummyDeployOpt
+            .traverse {
+              case (privateKey, term) =>
+                val deployData = ConstructDeploy.sourceDeployNow(
+                  source = term,
+                  sec = privateKey,
+                  vabn = nextBlockNum - 1,
+                  shardId = shardId
+                )
+                BlockDagStorage[F].addDeploy(deployData).as(List(deployData.sig))
+            }
+          OptionT
+            .fromOption(pooledOk.nonEmpty.guard[Option].as(pooledOk))
```

Comment:
`OptionT.whenF(pooledOk.nonEmpty)(pooledOk.pure)` looks more natural for me. @tgrospic, it such replacement correct?

Reply:
Yes, this also makes sense.

Another solution is to not even go to OptionT, but just use Option. In this case it's not necessary to mention `pooledOk` twice.
`Option(pooledOk).filter(_.nonEmpty)`